### PR TITLE
gh-1105 Make the Shell Role-aware

### PR DESCRIPTION
**Pull request by ghillert**
* Add role handling to methods annotated with `@CliAvailabilityIndicator`
* Add access handling rules to `DataFlowShell#hasAccess`
* Add tests
* Improve slightly handling of `403` (Forbidden) error responses in `VndErrorResponseErrorHandler`
  - See also https://github.com/spring-cloud/spring-cloud-dataflow/issues/1175

Resolves https://github.com/spring-cloud/spring-cloud-dataflow/issues/1105
Depends on https://github.com/spring-projects/spring-shell/issues/115

### CODE REVIEW

File: spring-cloud-dataflow-shell-core/src/main/java/org/springframework/cloud/dataflow/shell/config/DataFlowShell.java [hunk 1]
```diff
@@ -16,25 +16,125 @@
...
+	 * @param role Can be null.
+	 * @param opsType Must not be null.
+	 * @return If true the indicated operation is accessible and the user has all credentials to execute the operation.
+	 */
+	public boolean hasAccess(RoleType role, OpsType opsType) {
+		if (this.dataFlowOperations != null) {
+			boolean operationsAvailable = false;
+			switch(opsType) {
+				case AGGREGATE_COUNTER:
+					operationsAvailable = this.dataFlowOperations.aggregateCounterOperations() != null;
```
Comment (ericbottard) [inline spring-cloud-dataflow-shell-core/src/main/java/org/springframework/cloud/dataflow/shell/config/DataFlowShell.java]: I feel like all these "available" checks belong to the commands, where they were initially. Why not `&&` them in the `@CliAvailabilityIndicator` methods?

File: spring-cloud-dataflow-shell-core/src/main/java/org/springframework/cloud/dataflow/shell/command/ConfigCommands.java [hunk 2]
```diff
@@ -234,25 +245,15 @@ public String info() {
...
-		//Only invoke if the shell is executing in the same application context as the data flow server.
+	public void onApplicationEvent(ApplicationEvent event) {
 		if (!initialized) {
-			target(this.serverUri, this.userName, this.password, this.skipSslValidation);
-		}
-	}
-
-	@Override
-	public void afterPropertiesSet() throws Exception {
-		//Only invoke this lifecycle method if the shell is executing in stand-alone mode.
```
Comment (ericbottard) [inline spring-cloud-dataflow-shell-core/src/main/java/org/springframework/cloud/dataflow/shell/command/ConfigCommands.java]: I can't say I understood the rationale for the previous logic and the comment that went with it, but haven't the semantics changed with the commit you're proposing?

Comment (ericbottard) [inline spring-cloud-dataflow-shell-core/src/main/java/org/springframework/cloud/dataflow/shell/command/ConfigCommands.java]: My bad, I realize now that you even removed the `afterPropertiesSet ()` method. But then, isn't `initialized` always false?

File: spring-cloud-dataflow-shell-core/src/main/java/org/springframework/cloud/dataflow/shell/config/DataFlowShell.java (hunk 1 above)
Comment (ghillert) [inline spring-cloud-dataflow-shell-core/src/main/java/org/springframework/cloud/dataflow/shell/config/DataFlowShell.java]: I wanted to decouple the checking from the commands as much as possible for 2 reasons:

* Realizing that adding the access logic is quite repetitive, I wanted to make the code in the command-class look at simple as possible - 1 line:

```
return dataFlowShell.hasAccess(RoleType.VIEW, OpsType.AGGREGATE_COUNTER);
```

* ability to test the logic in insolation

File: spring-cloud-dataflow-shell-core/src/main/java/org/springframework/cloud/dataflow/shell/command/ConfigCommands.java (hunk 2 above)
Comment (ghillert) [inline spring-cloud-dataflow-shell-core/src/main/java/org/springframework/cloud/dataflow/shell/command/ConfigCommands.java]: Ooops, that is a hang-over from the previous PR. Wish that code would have been a bit cleaner. Will fix.

**Comment by ghillert** [comment]
@ericbottard  Made the changes to Config Command. Let me know if any further changes are needed.

**Comment by ericbottard** [comment]
Squashed and merged as c9f37779

**Pull request closed without merging**